### PR TITLE
Implement Operator::FormDiscreteOperator() [rectangular-parallel-operator]

### PR DESCRIPTION
**Pull request by barker29**
`Operator::FormLinearSystem()` assumes a square operator, but you may want to form a parallel version (on truedofs) of a rectangular operator (on dofs), for example for the matrix-free implementation of Nedelec interpolation and gradient operators. This PR implements some of the functionality of `Operator::FormLinearSystem()` in a new `Operator::FormParallelOperator()` method for rectangular operators.

### CODE REVIEW

File: linalg/operator.hpp [hunk 1]
```diff
@@ -113,6 +113,18 @@ class Operator
...
+       This has some of the functionality of Operator::FormLinearSystem(), but
+       works additionally on rectangular matrices.
+
+       The constraints are specified through the prolongation P from
+       GetProlongation(), and restriction R from GetRestriction() methods, which
+       are e.g. available through the (parallel) finite element space of any
+       (parallel) bilinear form operator. We have: `A(X)=[R (*this)
+       P](X)`. */
+   void FormParallelOperator(Operator* &A);
+
```
Comment (tzanio) [inline linalg/operator.hpp]: In the current `ParBilinearForm` this is called `FormSystemMatrix`:
```c++
/// Form the linear system matrix @a A, see FormLinearSystem() for details.
void FormSystemMatrix(const Array<int> &ess_tdof_list, OperatorHandle &A);
```

In the new `okina-integ` branch (see  #729) it is called `FormSystemOperator`:
```c++
/// Form the linear system operator @a A, see FormLinearSystem() for details.
virtual void FormSystemOperator(const Array<int> &ess_tdof_list,
                                Operator *&A);
```
I personally prefer the latter.

File: linalg/operator.hpp [hunk 2]
```diff
@@ -78,8 +78,8 @@ class Operator
        ParBilinearForm::FormLinearSystem()).
 
        The constraints are specified through the prolongation P from
-       GetProlongation(), and restriction R from GetRestriction() methods, which
-       are e.g. available through the (parallel) finite element space of any
+       GetProlongation(), and we assume the restriction is its transpose.
+       P is e.g. available through the (parallel) finite element space of any
        (parallel) bilinear form operator. We assume that the operator is square,
```
Comment (tzanio) [inline linalg/operator.hpp]: Good catch

File: linalg/operator.cpp [hunk 3]
```diff
@@ -69,6 +69,16 @@ void Operator::RecoverFEMSolution(const Vector &X, const Vector &b, Vector &x)
    }
 }
 
+void Operator::FormParallelOperator(Operator* &Aout)
```
Comment (tzanio) [inline linalg/operator.cpp]: Since this could be used for square operators, we probably should also add boundary conditions (or have two separate methods)

**Comment by tzanio** [comment]
Mention in `CHANGELOG`

**Comment by tzanio** [comment]
Hi Andrew,

Looking again at this, it seems to me we actually need two functions:

- `FormDiscreteOperator` (or similar name) that assumes a `DiscreteLinearOperator`-like rectangular mapping between two finite element spaces, fes1->fes2, without boundary conditions, and computes R2 G P1.

- `FormSystemOperator` that assumes a `BilinearForm`-like square operator, with boundary conditions, and computes P^T A P

I don't think the current implementation actually performs the first operation: we really need to get the restriction from the Nedelec space and the interpolation from the nodal space...

Am I missing something?

Tzanio

**Comment by barker29** [comment]
I've tried to address your comments. I am not entirely sure a square `FormSystemOperator()` is necessary because it duplicates a lot of what is in `FormLinearSystem()`, but it is there now.

Regarding your last point: the user/implementer needs to have their own `GetRestriction()` and `GetProlongation()` implemented, for my H(curl) implementation these do indeed come from the Nedelec and Lagrange spaces respectively. This is no different from what was previously the case for `Operator::FormLinearSystem()` which needed eg. `TBilinearForm` to have implemented the proper parallel prologation operator.

**Comment by tzanio** [comment]
Hi Andrew,

The issue is that `GetRestriction()` is for input vectors (in particular it is used to restrict the initial guess in `FormLinearSystem`). I added another method for output vectors, `GetOutputRestriction()` to keep those concepts separate.

Thanks!
Tzanio

**Pull request merged**